### ISSUE TITLE
Issue/Question: Path of the Spotify logo used in the panel

### ISSUE THREAD

**Issue by illegalmushroom** [esenliyim/sp-tray#26]
Hi, I really appreciate the work you've done on this extension, but I have a question regarding the Spotify logo option. Where is the path to the icon used by the extension, because I want to modify it and hopefully fix the issue I'm having (seen below). Thanks in advance!
![image](https://user-images.githubusercontent.com/88538398/168305634-1a18eb3c-a6f0-4bb7-84b2-1ed9d6ac98d4.png)

**Comment by esenliyim** [comment]
That feature was introduced by #20. Not my creation, and admittedly one of the things I need to take a closer look at and understand better. 

That problem occurs when Spotify is installed as a containerized/sandboxed package, like Snap or AppImage or Flatpak, which I'm guessing is how you have installed it on your machine. 

I haven't done any real investigating yet, but my educated guess is that the icon comes from wherever your system stores application icons by default, you know, where desktop entries and whatnot get their icons from. That'd explain why the icon is not present when the Spotify client is running in a sandbox. Missing application icons is unfortunately an annoyingly common problem with snap/flatpak/appimage packages. One solution to that would be packaging the icon with the extension so it doesn't have to rely on the system to provide it.

**Comment by illegalmushroom** [comment]
You're right, my Spotify is indeed installed as a Flatpak. I'll try to see if I can get it to play nicely, in the meantime it's up to you if this issue should remain open or not. Thanks for the clarification!

**Comment by e7d** [comment]
Hi @esenliyim and @illegalmushroom,

Having this issue since i upgraded from Fedora 35 to Fedora 36, I investigated a bit this morning, and it seems somethings change in the way icons are handled. And i guess Gnome 42 is the culprit here.

From what i reckon, the icon is loaded here:
https://github.com/esenliyim/sp-tray/blob/master/panelButton.js#L142
```
this.ui.set("icon", new St.Icon({
    icon_name: "spotify",
    style_class: "system-status-icon",
}))
```

As @esenliyim expressed himself, loading icons this way loads them from the system using a name as a reference.
This icon catalog can actually be consulted using the app "GTK Icon Browser".
On my Fedora 36, the binary for the latest GTK version (GTK4) can be installed by fetching the `gtk4-devel-tools` package, giving you the `gtk4-icon-browser` binary amongst others.

Launching it and browsing a bit, you can see the `spotify` icon does not exist (anymore?).
![image](https://user-images.githubusercontent.com/1412033/169254462-9f1b81ec-3d5c-439d-a160-0d541d4284f8.png)

As @esenliyim said:
> One solution to that would be packaging the icon with the extension so it doesn't have to rely on the system to provide it.
But what about copyright, at least with the official one? :shrug: 

The say here https://developer.spotify.com/design/content/ :
![image](https://user-images.githubusercontent.com/1412033/169255053-a9dac05d-f535-4630-b77a-c915025b480b.png)
More details here, but did not read it all:
https://developer.spotify.com/terms/

Maybe asking them kindly on their community forum would suffice after all, arguing this is open-source software?

In any case, I gave it a try [on this branch](https://github.com/e7d/sp-tray/commit/e58ddb2b57e143c0cf83d2548a6076779c33f40e), with nice results:
![image](https://user-images.githubusercontent.com/1412033/169265147-caefc28d-cebd-4471-9467-1f62b451b412.png)
![image](https://user-images.githubusercontent.com/1412033/169265404-68efb438-b60c-4cd3-846c-f6958df24a29.png)

**Comment by e7d** [comment]
> Maybe asking them kindly on their community forum would suffice after all, arguing this is open-source software?

No need to go on their forum, the link "approved by Spotify" we can see on the screenshot is actually a form to ask for approbation.
It directs here: https://spotify.atlassian.net/servicedesk/customer/portal/3/create/55

**Comment by esenliyim** [comment]
Good write-up.

So just for clarification, were you using a flatpak/snap/whatever package of Spotify on Fedora 35 too? And the logo worked then?

As for the GTK4 icon browser here, that thing is unfortunately telling you nothing. It doesn't show the Spotify icon on my system either, but there it is, being displayed just fine by the extension:

![image](https://user-images.githubusercontent.com/33372147/169282093-cc4fd42c-1776-48f7-afad-1e9ee7a269c8.png)

I'll look into the permission and #28 when I have the time.

**Comment by e7d** [comment]
I'm using the flatpak version now, but I know I did switch from a dnf installed package to the flatpak version at some point, while I was still using Fedora 35. And i can not quite remember if the logo still was showing after that switch. :thinking: 
I'll try in a brand new Fedora 35 VM and let you know.

**Comment by esenliyim** [comment]
The logo is a relatively new feature. April 7th, according to #20. If you've made the switch before that, you've always used it with a containerized package.

**Comment by e7d** [comment]
Well, I took the time to test again both installation methods. They are listed here btw: https://docs.fedoraproject.org/en-US/quick-docs/installing-spotify/#_using_the_rpm_fusion_repository

I started by installing both packages side by side, as their disk location are completely different.
The Spotify does not show, whatever the version started, even having only one of them running at a time, in case running both of them concurrently could mess up the detection for whatever reason.

By the way, the metadata of your plugin always reacts correctly to this "concurrency situation". Whatever the client used, the other one, as well as your plugin syncs up right away with the good information.
Nice to see.

Following your previous comment, I then uninstalled the flatpak version completely, and restarted the VM for good measure, so the one installed through "local package factory" would be the only one in place.
Still no logo.

Finally, and to be thorough, I did the opposite and removed any trace of the "lpf-spotify-client" and its build "spotify-client" packages. I installed back the Flatpak package, and restarted the VM.
No logo again.

I don't know if my memory is playing tricks on me, but I'm pretty sure I saw this Spotify logo show up some day, and yet i tried every combination of what I used past months.
I shall say I'm quite confused now.

**Comment by esenliyim** [comment]
I know this is gonna sound like "did you turn it off and back on again", but I'm gonna ask it anyway just to make sure. 

You did remember to actually set the logo to be shown, right? Because a few commits back I've set it to be hidden by default.

**Comment by esenliyim** [comment]
Solved this one too, at least partially. Turns out the app icon of the flatpak build is very easy to get. You get it the same way you do the "normal" icon, it's just the name that's different. Adding a single line of code did it.

Snap isn't that easy though, and I haven't tried with AppImage as I believe there isn't an official build for it. Leaving the issue open until I can get it working with Snap too.

**Comment by esenliyim** [comment]
Okay I'm getting close with Snap as well.

The solution I currently have is to check for Spotify's `.desktop` file where snapd stores those. If it exists, then I read the location of the icon off of it. I feel it's a rather overengineered solution, but I can't bake the icon's absolute path into the code and go straight for it, because the path includes what I'm pretty sure is a version number that changes from version to version. That'd break with the next update. 

Maybe I'm dumb or haven't looked into it enough, but that's the best I got at the moment.

**Comment by e7d** [comment]
> Solved this one too, at least partially. Turns out the app icon of the flatpak build is very easy to get. You get it the same way you do the "normal" icon, it's just the name that's different. Adding a single line of code did it.
> 
> Snap isn't that easy though, and I haven't tried with AppImage as I believe there isn't an official build for it. Leaving the issue open until I can get it working with Snap too.

That did the trick for my flatpak installation, thanks.

**Comment by brw** [comment]
I know this is a bit of an old issue, but I noticed that the Spotify icon also isn't shown for my installation. I think it might be because I'm running Spotify through [spotify-launcher](https://archlinux.org/packages/extra/x86_64/spotify-launcher) (as recommended by the [Arch wiki](https://wiki.archlinux.org/title/spotify), and it makes it easier to work with [Spicetify](https://github.com/spicetify/spicetify-cli)), which installs Spotify to `$HOME/.local/share/spotify-launcher/install`. The global .desktop file is `/usr/share/applications/spotify-launcher.desktop` instead of `spotify.desktop` (which is located at `$HOME/.local/share/spotify-launcher/install/usr/share/spotify/spotify.desktop`, but unused). Was just wondering if this is also the same issue.

![image](https://github.com/esenliyim/sp-tray/assets/7560938/c7f28a4a-4555-4ffd-b297-2d6a8fd4917a)
